1_dynamic/future.k

```diff
@@ -16,4 +18,465 @@ endmodule
+  * the (already evaluated) arguments to be passed to the method.
+*/
+
+  // the arguments are strict such that they are evaluated first
+  context _:ObjectVal ! Method:Id(HOLE)
+
+  rule <k> object(OId,Class,_) ! Method:Id(Vs:Vals)
+        => future(F) ...</k>
+       (. => <future>...
+               <fid> F </fid> <oid> OId </oid>
+               <closure> lookupMethod(Class,Method) </closure>
+               <arguments> Vs </arguments>
+             ...</future>)
+       <nextFutureId> F:Int => F +Int 1 </nextFutureId>
+
+  // future identifiers
+  syntax Val ::= future(Int)
+  syntax KItem ::= lookupMethod(Id,Id)
+
+/* Looking up the location of a future method by name
+  * The method is looked up in the envStack of the object,
+    starting from specified class
+  * can be done in parallel with any existing tasks
+*/
+  rule <future>
+         <fid> F:Int </fid>
+         <oid> Id:Int </oid>
+          <closure>
+             lookupMethod(Class,Method)
+          => lookupThis(<envStack> ES </envStack>, Class, Method)
+          </closure>
+         <arguments> Vs:Vals </arguments>
+         <state> sleeping </state>
+       </future>
+       <location> Id </location>
+       <envStack> ES:List </envStack>
+
+/* 
+  The activation of a sleeping future
+  * a future can only be activated if there is no active state
+    - signaled by futureId holding -1
+  * performCall does the actual method invocation
+*/
+
+  rule <future> 
+         <fid> F:Int </fid>
+         <oid> Id:Int </oid>
+         <closure> Closure </closure>
+         <arguments> Vs:Vals </arguments>
+         <state> sleeping => active </state>
+       </future>
+       <location> Id </location>
+       <envStack> ES:List </envStack>
+       <futureId> -1 => F </futureId>
+       <threads>
+         .
+         =>
+         <thread>...  <k> performCall(Closure, Vs) </k> ...</thread>
+       </threads>
+    [futureStart]
+
+/*
+  Performing a local call:
+  * Parameters are bound to arguments, then the body is executed
+  * the current reference class is updated to the class of the method
+*/
+  syntax KItem ::= performCall(K, Vals) [strict(1)]
+
+  rule <k> performCall(methodClosure(_,Class,OL,Ps,S), Vs:Vals)
+           => mkDecls(Ps,Vs) S
+       </k>
+       <crntClass> _ => Class </crntClass>
+
+
+/*  Semantics for the class initializer. See definition of KOOL for create.
+*/
+  rule <k> performCall($clinitClosure, _) => create(Class) </k>
+       <baseClass> Class:Id </baseClass>
+       <envStack> .List => ListItem((Object, <env>.Map</env>)) </envStack>
+
+/*
+  Method calls desugar into (blocking) future invocations
+*/
+
+  rule O:ObjectVal . Method:Id(Vs:Vals) 
+    => (O ! Method(Vs)).get
+  
+/*
+  A super call desugars into a normal call using a fake object
+  having the super class as a base.
+*/
+
+  rule <k> super . Method:Id(Vs:Vals)
+        => object(Id, Class', Class') . Method(Vs)
+       ...</k>
+      <crntClass> Class </crntClass>
```

Comment:
Well, we're not actually calling the method here;  it will be changed when the method is called in the future task.